projects/odf_transform/odf_transform/odf_parser/odf_parser.py

```diff
@@ -40,275 +50,383 @@ def read(filename,
+                "gf3_code"
+            ):
+                warnings.warn(
+                    "{0}[{1}] flag was matched to the variable {2} but ODF attributes {3} do not match".format(
+                        var,
+                        att["long_name"],
+                        related_variable,
+                        {
+                            key: metadata[related_variable].get(key)
+                            for key in ["long_name", "gf3_code"]
+                        },
+                    ),
+                    UserWarning,
+                )
+
+            # Standardize long name attribute of flag variables
+            if "name" in metadata[related_variable]:
+                att["long_name"] = (
+                    flag_long_name_prefix + metadata[related_variable]["name"]
+                )
             else:
-                previous_key = i
-        return previous_key
+                att["long_name"] = flag_long_name_prefix + related_variable
 
-    # Find matching vocabulary
+            # Add ancillary_variables attribute
+            if "ancillary_variables" in metadata[related_variable]:
+                metadata[related_variable]["ancillary_variables"] += ",{0}".format(var)
+            else:
+                metadata[related_variable]["ancillary_variables"] = var
 
-    flag_dict = {}
-    for var, att in metadata.items():
+        # Add flag convention attributes if available within config file
+        if flag_convention:
+            if var in flag_convention:
+                att.update(flag_convention[var])
+            elif "default" in flag_convention:
+                att.update(flag_convention["default"])
+
+        # TODO rename QQQQ_XX flag variables to Q[related_variables] so that ERDDAP can easily amalgamate them!
+    return metadata
+
+
+def get_vocabulary_attributes(
+    metadata, organizations=None, vocabulary=None, vocabulary_attribute_list=None
+):
+    """
+    This method is use to retrieve from an ODF variable code, units and units, matching vocabulary terms available.
+    """
+
+    def _compare_units(unit, expected_units):
+        """Simple tool to compare "|" separated units in the Vocabulary expected unit list.
+        - First unit if any is matching.
+        - None if empty or expected to be empty
+        - False if not matching units
+        """
+        if expected_units:
+            # Split unit list and convert None or dimensionless to None
+            unit_list = []
+            for unit_item in expected_units.split("|"):
+                if unit_item.lower() in ["none", "dimensionless"]:
+                    unit_list.append(None)
+                else:
+                    unit_list.append(unit_item)
+
+            standard_unit = unit_list[0]
+        else:
+            unit_list = []
+            standard_unit = None
+
+        if unit in ["none", "dimensionaless"]:
+            unit = None
+
+        if unit in unit_list:
```

Comment:
theres a lot of logic here, would be nice to have comments about what it all means